File: domain/src/test/java/com/kaciula/archiman/domain/usecases/GetUsersTest.java
package com.kaciula.archiman.domain.usecases;


import com.kaciula.archiman.domain.boundary.data.UserRepository;
import com.kaciula.archiman.domain.entity.User;
import com.kaciula.archiman.domain.util.TrampolineSchedulerProvider;
import io.reactivex.Observable;
import io.reactivex.observers.TestObserver;
import java.util.ArrayList;
import java.util.List;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;

public class GetUsersTest {

  @Mock UserRepository userRepository;

  private GetUsers useCase;

  @Before
  public void setupUseCase() {
    MockitoAnnotations.initMocks(this);

    useCase = new GetUsers(userRepository, new TrampolineSchedulerProvider());
  }

  @Test
  public void getUsersSuccess() {
    final List<User> users = new ArrayList<>();
    users.add(User.create(1, "Best programmer"));
    users.add(User.create(2, "Second best programmer"));
    Mockito.when(userRepository.getUsers()).thenReturn(Observable.fromArray(users));

    TestObserver<GetUsers.ResponseModel> observer =
        useCase.execute(GetUsers.RequestModel.create()).test();
    observer.assertValues(GetUsers.ResponseModel.IN_FLIGHT, GetUsers.ResponseModel.success(users));
  }

  @Test
  public void getUsersFailure() {
    Throwable error = new Throwable();
    Mockito.when(userRepository.getUsers()).thenReturn(Observable.error(error));

    TestObserver<GetUsers.ResponseModel> observer =
        useCase.execute(GetUsers.RequestModel.create()).test();
    observer.assertValues(GetUsers.ResponseModel.IN_FLIGHT, GetUsers.ResponseModel.error(error));
  }
}

File: domain/src/test/kotlin/com/kaciula/archiman/domain/usecases/GetUsersTest.kt
package com.kaciula.archiman.domain.usecases


import com.kaciula.archiman.domain.boundary.infrastructure.UserRepository
import com.kaciula.archiman.domain.model.User
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations

class GetUsersTest {

    @Mock private lateinit var userRepository: UserRepository

    private lateinit var useCase: GetUsers

    @Before
    fun setupUseCase() {
        MockitoAnnotations.initMocks(this)

        useCase = GetUsers(userRepository)
    }

    @Test
    fun getUsersSuccess() {
        val users = arrayListOf(User(1, "Best programmer"), User(2, "Second best programmer"))
        Mockito.`when`(userRepository.getAll()).thenReturn(Single.just(users))

        val observer = useCase.execute(GetUsers.RequestModel).test()
        observer.assertValue(GetUsers.ResponseModel(users = users))
    }

    @Test
    fun getUsersFailure() {
        val error = Throwable()
        Mockito.`when`(userRepository.getAll()).thenReturn(Single.error(error))

        val observer = useCase.execute(GetUsers.RequestModel).test()
        observer.assertError(error.javaClass)
    }
}
